### PR TITLE
Add post on scaling to 15k functions

### PR DESCRIPTION
**Pull request by alexellis**
## Description

Add post on scaling to 15k functions

## Motivation and Context

Explain/educate/inspire other vendors and customers about what can be done.

Show our working, and innovation.

### CODE REVIEW

File: _posts/2023-10-19-large-scale-functions.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+**What does it cost to test at scale?**
+
+In addition, with OpenFaaS running on top of Kubernetes, we have to provision a whole node for every 100-110 containers that are provisioned, including control-plane, service mesh and networking. So for 3000 functions, you need at least 30 nodes.
+
+We all know that clusters are slow to create on a platform like AWS EKS, and then adding nodes can take a good 3-5 minutes each. I did a calculation with the AWS cost estimator and to get to 3500 functions, you'd probably be looking at a spend of 1500 USD / mo in infrastructure costs alone.
+
+**How did we find the problem?**
+
+The problem was finally found after spending after building over a week of optimizations, and it was frustratingly obvious.
```
Comment (welteki) [inline _posts/2023-10-19-large-scale-functions.md:36]: This sentence has a weird structure. You intention was probably to write something like:

```suggestion
The problem was finally found after spending over a week building  optimizations, and it was frustratingly obvious.
```

File: _posts/2023-10-19-large-scale-functions.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+That overwhelmed the server when I ran all 120 `k3sup join` commands at once, so now `k3sup node-token` will get the token, either into a file or into memory, and can then be passed in via `k3sup join --node-token`.
+
+## What did we change in OpenFaaS?
+
+**Leader election**
+
+I was in two minds to implement lease-based Leader Election), because it's a divisive topic. Some people haven't had any issues, but others have had significant downtime and have experienced extra load on the API server due using it.
+
+![Lease-based leader election](images/2023-10-scale-operator/reconcile.png)
```
Comment (welteki) [inline _posts/2023-10-19-large-scale-functions.md:116]: Image is not loading due to missing `/`:
```suggestion
![Lease-based leader election](/images/2023-10-scale-operator/reconcile.png)
```

File: _posts/2023-10-19-large-scale-functions.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+> Lease-based leader election
+
+When three replicas of the Gateway Pod start up, each starts a REST API which can serve invocations, and the REST API for configuring Functions. However, only one of the three replicas should be performing reconciliation of Functions into Deployments and Services, so whichever takes the lease will act as a leader, and the others will just stand by. If the leader gives up the lease due to a graceful shutdown, another will take over. If the crashes or a spot instance is terminated, then the lease will expire after 60s and another replica will take over.
+
+Leader Election is optional and disabled by default, but if you are running more than replica of the gateway, it's recommended, and prevents noise from conflicting writes or updates, which must in turn be evaluated by the Kubernetes API server and OpenFaaS.
```
Comment (welteki) [inline _posts/2023-10-19-large-scale-functions.md:121]:
```suggestion
Leader Election is optional and disabled by default, but if you are running more than one replica of the gateway, it's recommended, and prevents noise from conflicting writes or updates, which must in turn be evaluated by the Kubernetes API server and OpenFaaS.
```

File: _posts/2023-10-19-large-scale-functions.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+Waiting for caches to sync for faas-netes:Service
+```
+
+**Many log messages have been removed**
+
+We took the verbosity down a level.
+
+Previously, if a Function CR had been created, but not yet reconciled, then the logs for the operator would have printed a message saying the Deployment was not available, whenever a component tried to list functions. That was noise that we just didn't need so it was taken away.
+
+The same is the case for when a function is deployed via REST API, we used to print out a message saying "Deploying function X". Well, that's very noisy when you are trying to create 150000 functions in a short period of time.
```
Comment (welteki) [inline _posts/2023-10-19-large-scale-functions.md:191]: 150000 or 15000 functions?

File: _posts/2023-10-19-large-scale-functions.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+$ kubectl scale deploy/nodeinfo openfaas-fn --replicas=5
+
+# View the endpoints
+$ kubectl get endpoints/nodeinfo -n openfaas-fn -o yaml
+
+# View the slices, you should see one:
+$ kubectl get endpointslice -n openfaas-fn | grep nodeinfo
+nodeinfo-9ngtv   IPv4          8080    10.244.0.165   8s
+
+$ kubectl get endpointslice/nodeinfo-9ngtv -n openfaas-fn -o yaml
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2023-10-19-large-scale-functions.md:169]: It's an exercise for the reader..

**Pull request merged**